python/tests/test_pipeline_builder.py

```diff
@@ -647,6 +649,175 @@ def test_input_json1(self):
 
         pipeline.delete()
 
+    def test_issue2142(self):
+        sql = f"""
+        CREATE TABLE t0 (c1 INT);
+        CREATE VIEW v0 AS SELECT * FROM t0;
+        """
+
+        pipeline = PipelineBuilder(TEST_CLIENT, name="test_issue2142", sql=sql).create_or_replace()
+
+        data = [{"c1": None}, {"c1": 1}]
+
+        out = pipeline.listen("v0")
+
+        pipeline.start()
+        pipeline.input_json("t0", data=data)
+        pipeline.wait_for_completion(True)
+
+        out_data = out.to_dict()
+
+        for datum in data:
+            datum.update({"insert_delete": 1})
+
+        assert out_data == data
+
+        pipeline.delete()
+
+    def test_pandas_binary(self):
+        sql = f"""
+        CREATE TABLE t0 (c1 VARBINARY);
+        CREATE VIEW v0 AS SELECT SUBSTRING(c1 FROM 2) as c1 FROM t0;
+        """
+
+        data = [{"c1": [12, 34, 56]}]
+        expected_data = [{"c1": [34, 56], "insert_delete": 1}]
+
+        pipeline = PipelineBuilder(TEST_CLIENT, name="test_pandas_binary", sql=sql).create_or_replace()
+        out = pipeline.listen("v0")
+
+        pipeline.start()
+        pipeline.input_json("t0", data=data)
+        pipeline.wait_for_completion(True)
+
+        got = out.to_dict()
+
+        assert expected_data == got
+
+    def test_pandas_decimal(self):
+        from decimal import Decimal
+
+        sql = f"""
+        CREATE TABLE t0 (c1 DECIMAL(5, 2));
+        CREATE VIEW v0 AS SELECT c1 + 2.75::DECIMAL(5, 2) as c1 FROM t0;
+        """
+
+        data = [{"c1": 2.25}]
+        expected = [{"c1": Decimal('5.00'), "insert_delete": 1}]
+
+        pipeline = PipelineBuilder(TEST_CLIENT, name="test_pandas_decimal", sql=sql).create_or_replace()
+        out = pipeline.listen("v0")
+
+        pipeline.start()
+        pipeline.input_json("t0", data=data)
+        pipeline.wait_for_completion(True)
+
+        got = out.to_dict()
+
+        assert expected == got
+
+    def test_pandas_array(self):
+        sql = f"""
+        CREATE TABLE t0 (c1 INT ARRAY);
+        CREATE VIEW v0 AS SELECT c1 FROM t0;
+        """
+
+        data = [{"c1": [1, 2, 3]}]
```

Comment:
So, it falls back to doing type inference for arrays, maps, and structs or do they end up being dynamically typed dictionaries?

Reply:
Well, pandas just treats it as `object`. 

Basically the same as if there were no schema, they are just list / dict objects. So, dynamically typed dictionaries and lists.

Reply:
So this means that the same primitive types end up having different representations when they are stored in a column vs in an array/struct/map. I guess we can leave this for future work, but this should be somehow documented.